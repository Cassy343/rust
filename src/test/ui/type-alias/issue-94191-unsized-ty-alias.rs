// Check for improved type alias diagnostics

struct Container<T>(T);

type UnsizedContainer = Container<str>;
type ContainerAlias<T> = Container<T>;

fn takes_container(_: UnsizedContainer) {}
//~^ ERROR the size for values of type `str` cannot be known at compilation time [E0277]

fn takes_generic_container<T>(_: ContainerAlias<T>)
//~^ ERROR the size for values of type `T` cannot be known at compilation time [E0277]
where
    T: ?Sized,
{
}

fn main() {}
