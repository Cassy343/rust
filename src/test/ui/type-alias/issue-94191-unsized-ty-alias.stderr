error[E0277]: the size for values of type `str` cannot be known at compilation time
  --> $DIR/issue-94191-unsized-ty-alias.rs:8:23
   |
LL | fn takes_container(_: UnsizedContainer) {}
   |                       ^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `str`
note: required by a bound in `Container`
  --> $DIR/issue-94191-unsized-ty-alias.rs:3:18
   |
LL | struct Container<T>(T);
   |                  ^ required by this bound in `Container`
help: you could relax the implicit `Sized` bound on `T` if it were used through indirection like `&T` or `Box<T>`
  --> $DIR/issue-94191-unsized-ty-alias.rs:3:18
   |
LL | struct Container<T>(T);
   |                  ^  - ...if indirection were used here: `Box<T>`
   |                  |
   |                  this could be changed to `T: ?Sized`...

error[E0277]: the size for values of type `T` cannot be known at compilation time
  --> $DIR/issue-94191-unsized-ty-alias.rs:11:34
   |
LL | fn takes_generic_container<T>(_: ContainerAlias<T>)
   |                            -     ^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   |                            |
   |                            this type parameter needs to be `std::marker::Sized`
   |
note: required by a bound in `Container`
  --> $DIR/issue-94191-unsized-ty-alias.rs:3:18
   |
LL | struct Container<T>(T);
   |                  ^ required by this bound in `Container`
help: you could relax the implicit `Sized` bound on `T` if it were used through indirection like `&T` or `Box<T>`
  --> $DIR/issue-94191-unsized-ty-alias.rs:3:18
   |
LL | struct Container<T>(T);
   |                  ^  - ...if indirection were used here: `Box<T>`
   |                  |
   |                  this could be changed to `T: ?Sized`...
help: consider removing the `?Sized` bound to make the type parameter `Sized`
   |
LL - where
LL -     T: ?Sized,
   |

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
